### PR TITLE
Accelerate Parquet en/decryption with HTTPFS extension

### PR DESCRIPTION
**Pull request by ccfelius**
This PR enables to accelerate Parquet en/decryption with OpenSSL when the httpfs extension is loaded.

TL;DR Parquet en/decryption uses the mbedtls library  by default. However, when using the httpfs extension we can leverage OpenSSL to speedup the en/decryption process. The en/decryption of Parquet files will be accelerated by default. This means that it _does not require any additional steps_ to enable the acceleration once httpfs is build.

### How it works
---
When the httpfs extension is loaded, a pointer (`EncryptionUtil`) to a factory method (`AESGCMStateSSLFactory`) will be set in the `DBconfig`. 

If a Parquet file needs to be encrypted or decrypted, the Parquet reader/writer checks whether the `EncryptionUtil` pointer is set upon initialization. If the statement is false, an `AESGCMStateMBEDTLSFactory`  pointer will be set; otherwise an `AESGCMStateSSLFactory` pointer from `DBconfig` will be used. 

When an actual call to  `LoadMetadata`, `ParquetReader::Read`,  `ParquetReader::ReadData`, `ParquetWriter::Write` or `ParquetWriter::WriteData` occurs (and the `encryption_config` is present) an actual object will be created by `CreateEncryptionState()`. This object represents the Encryption (AES) state. For Parquet, the state can either consist of an `AESGCMStateMBEDTLS` object, which uses mbedtls, or it can create an `AESGCMStateOpenSSL` object, which uses OpenSSL.

I have tried to maintain the same logic as used for the mbedtls-based en/decryption functions. However, some minor refactoring was required since the en/decryption logic for mbedtls and openssl slightly differs.

### Benchmarks
---
The benchmarks are performed on a Mac M2 Pro, using TPC-H with scale factor 1. The execution times are the average of 5 subsequent runs. I've added the exact benchmarks I used.

| | Unencrypted | OpenSSL | mbedtls
| -------- | ------- |  ------- |  ------- | 
| **Load  (TPCH)** | 0.271  | 0.294  | 0.503  |
| Overhead (Load) |   | _~9%_  | _~86%_  |
| **Q1 (TPCH)** | 0.070     |  0.075  |  0.153  |
| Overhead (Q1) |   | _~7%_  | _~119%_  | 

The overhead is reduced from _86%_ to _9%_ and from _119%_ to _7%_ for loading + encrypting data and performing Q1 on encrypted Parquet respectively. This difference is significant.

### Tests
---
I've added 3 new test files that repeat most of the existing tests for Parquet encryption:

- `parquet_encryption_httpfs.test` which is similar to `parquet_encryption.test` but then with httpfs enabled
- `parquet_encrypted_tpch_httpfs.test_slow`, which is similar to `parquet_encrypted_tpch.test_slow` with httpfs enabled
- `parquet_encryption_mbedtls_openssl.test` which tests the decrypting of files, that are encrypted with mbedtls, with OpenSSL and vice versa.

### Debugging with mbedtls when httpfs is build
---
If the httpfs extension is found, by default the application switches to OpenSSL encryption at runtime. However, if you want to debug with mbedtls enabled you can set  parquet option: `debug_use_openssl = false`. The final logic looks like this:

For copy:
```
# write files with mbedtls (explicitly disable OpenSSL if httpfs is loaded)
COPY (SELECT 42 i) to 'test_file.parquet' (ENCRYPTION_CONFIG {footer_key: 'key${key_len}'}, 
DEBUG_USE_OPENSSL false)
```

For read:
```
# read encrypted files by using mbedtls (explicitly disable OpenSSL if httpfs is loaded)
SELECT * FROM read_parquet('test_file.parquet', encryption_config={footer_key: 'key${key_len}'}, 
debug_use_openssl=false)
```

### CODE REVIEW

**Comment by Tishj** [comment]
Exciting improvements from the benchmarks you shared!

I think we'd be interested in adding a micro benchmark to catch any regressions in the future, these would be placed under `./benchmark/micro`

Using the [Benchmark Suite](https://duckdb.org/docs/dev/benchmark)

File: src/include/duckdb/main/config.hpp [hunk 1]
```diff
@@ -246,6 +249,8 @@ struct DBConfig {
 	shared_ptr<BufferManager> buffer_manager;
 	//! Set of callbacks that can be installed by extensions
 	vector<unique_ptr<ExtensionCallback>> extension_callbacks;
+	//! AES state for en/decryption
+	shared_ptr<AESStateFactory> encryption_state;
```
Comment (Mytherin) [inline src/include/duckdb/main/config.hpp:253]: Perhaps we can rename `AESStateFactory` to e.g. `EncryptionUtil` to be more generic/allow for additional expansion here in the future?

File: src/include/duckdb/main/config.hpp [hunk 2]
```diff
@@ -102,6 +103,8 @@ struct DBConfigOptions {
 	bool use_direct_io = false;
 	//! Whether extensions should be loaded on start-up
 	bool load_extensions = true;
+	//! Force to use mbedtls for encryption
+	bool parquet_use_openssl = false;
```
Comment (Mytherin) [inline src/include/duckdb/main/config.hpp:107]: Can we define this as an option in the Parquet extension instead of in the DBConfig?

File: extension/parquet/parquet_crypto.cpp [hunk 3]
```diff
@@ -195,19 +191,31 @@ class DecryptionTransport : public TTransport {
...
+			throw InternalException("DecryptionTransport::Finalize was called with bytes remaining in read buffer: \n"
+			                        "read buffer offset: %d, read buffer size: %d",
+			                        read_buffer_offset, read_buffer_size);
 		}
 
 		data_t computed_tag[ParquetCrypto::TAG_BYTES];
-		if (aes.Finalize(read_buffer, AESGCMState::BLOCK_SIZE, computed_tag, ParquetCrypto::TAG_BYTES) != 0) {
-			throw InternalException("DecryptionTransport::Finalize was called with bytes remaining in AES context");
+
+		if (aes->IsOpenSSL()) {
```
Comment (Mytherin) [inline extension/parquet/parquet_crypto.cpp:203]: Is there no way of unifying this in the callbacks? That would be cleaner as we can keep the logic in a single location

File: extension/parquet/parquet_crypto.cpp [hunk 4]
```diff
@@ -116,47 +110,50 @@ class EncryptionTransport : public TTransport {
...
 
 private:
 	//! Protocol and corresponding transport that we're wrapping
 	TProtocol &prot;
 	TTransport &trans;
 
-	//! AES context
-	AESGCMState aes;
+	//! AES context and buffers
+	shared_ptr<AESGCMState> aes;
```
Comment (Mytherin) [inline extension/parquet/parquet_crypto.cpp:156]: I realize this was here before as well - but Parquet does support [different encryption algorithms](https://github.com/apache/parquet-format/blob/master/Encryption.md#41-aes-modes-used-in-parquet). Perhaps we can call this encryption data as this might be extended in the future?

File: src/include/duckdb/main/settings.hpp [hunk 5]
```diff
@@ -481,6 +481,15 @@ struct MaximumTempDirectorySize {
 	static Value GetSetting(const ClientContext &context);
 };
 
+struct OpenSSLSetting {
+	static constexpr const char *Name = "parquet_use_openssl";
```
Comment (Mytherin) [inline src/include/duckdb/main/settings.hpp:485]: As mentioned above as well can this be moved into the Parquet extension?

File: extension/httpfs/include/crypto.hpp [hunk 6]
```diff
@@ -16,4 +17,42 @@ void hmac256(std::string message, hash_bytes secret, hash_bytes &out);
...
+	DUCKDB_API void GenerateRandomData(duckdb::data_ptr_t data, duckdb::idx_t len) override;
+
+public:
+	static constexpr size_t BLOCK_SIZE = 16;
+	static constexpr bool SSL = true;
+
+private:
+	void *gcm_context;
+	// 0 = encrypt, 1 = decrypt
+	bool mode;
```
Comment (Mytherin) [inline extension/httpfs/include/crypto.hpp:45]: Can this be an enum instead?

File: extension/httpfs/httpfs_extension.cpp [hunk 7]
```diff
@@ -59,6 +60,10 @@ static void LoadInternal(DatabaseInstance &instance) {
 	provider->SetAll();
 
 	CreateS3SecretFunctions::Register(instance);
+
+	// set parquet_use_openssl to true when httpfs is loaded
+	config.options.parquet_use_openssl = true;
```
Comment (Mytherin) [inline extension/httpfs/httpfs_extension.cpp:65]: Instead of setting this option here - perhaps Parquet can by default use `config.encryption_state` (i.e. have this setting default to true, where it tries to use `config.encryption_state` if available, and only when explicitly disabled do we not use that)?

File: extension/httpfs/crypto.cpp [hunk 8]
```diff
@@ -24,4 +30,118 @@ void hex256(hash_bytes &in, hash_str &out) {
...
+	// generate random bytes for nonce
+	RAND_bytes(data, len);
+}
+
+void AESGCMStateSSL::InitializeEncryption(const_data_ptr_t iv, idx_t iv_len, const string *key) {
+	auto context = reinterpret_cast<evp_cipher_ctx_st *>(gcm_context);
+	// Encryption mode
+	mode = false;
+	if (1 !=
+	    EVP_EncryptInit_ex(context, GetCipher(*key), NULL, reinterpret_cast<const unsigned char *>(key->data()), iv)) {
```
Comment (Mytherin) [inline extension/httpfs/crypto.cpp:74]: nit: we can use `char_ptr_cast`  or `data_ptr_cast` here

File: extension/httpfs/include/crypto.hpp [hunk 9]
```diff
@@ -16,4 +17,42 @@ void hmac256(std::string message, hash_bytes secret, hash_bytes &out);
...
+	DUCKDB_API size_t Finalize(duckdb::data_ptr_t out, duckdb::idx_t out_len, duckdb::data_ptr_t tag,
+	                           duckdb::idx_t tag_len) override;
+	DUCKDB_API void GenerateRandomData(duckdb::data_ptr_t data, duckdb::idx_t len) override;
+
+public:
+	static constexpr size_t BLOCK_SIZE = 16;
+	static constexpr bool SSL = true;
+
+private:
+	void *gcm_context;
```
Comment (Mytherin) [inline extension/httpfs/include/crypto.hpp:43]: Does this need to be a `void*`?

File: extension/httpfs/include/crypto.hpp [hunk 10]
```diff
@@ -16,4 +17,42 @@ void hmac256(std::string message, hash_bytes secret, hash_bytes &out);
...
 
+class AESGCMStateSSL : public duckdb::AESGCMState {
+
+public:
+	DUCKDB_API explicit AESGCMStateSSL();
+	DUCKDB_API ~AESGCMStateSSL() override;
+
+public:
+	DUCKDB_API bool IsOpenSSL() override;
+	DUCKDB_API void InitializeEncryption(duckdb::const_data_ptr_t iv, duckdb::idx_t iv_len,
```
Comment (Mytherin) [inline extension/httpfs/include/crypto.hpp:28]: nit: since this is in the duckdb namespace the `duckdb::` identifiers are unnecessary

File: extension/httpfs/include/crypto.hpp [hunk 11]
```diff
@@ -16,4 +17,42 @@ void hmac256(std::string message, hash_bytes secret, hash_bytes &out);
...
+private:
+	void *gcm_context;
+	// 0 = encrypt, 1 = decrypt
+	bool mode;
+};
+
+class AESGCMStateSSLFactory : public duckdb::AESStateFactory {
+public:
+	duckdb::shared_ptr<duckdb::AESGCMState> CreateAesState() const override {
+		return duckdb::shared_ptr<AESGCMStateSSL>(new AESGCMStateSSL());
```
Comment (Mytherin) [inline extension/httpfs/include/crypto.hpp:51]: nit: prefer to use `make_shared`

File: extension/httpfs/crypto.cpp [hunk 12]
```diff
@@ -24,4 +30,118 @@ void hex256(hash_bytes &in, hash_str &out) {
...
+		                              reinterpret_cast<int *>(&out_len));
+		text_len += out_len;
+
+		if (ret > 0) {
+			// success
+			return text_len;
+
+		} else {
+			throw InvalidInputException("Computed AES tag differs from read AES tag, are you using the right key?");
+			return -1;
```
Comment (Mytherin) [inline extension/httpfs/crypto.cpp:142]: nit: this code cannot be triggered

File: extension/httpfs/crypto.cpp [hunk 13]
```diff
@@ -24,4 +30,118 @@ void hex256(hash_bytes &in, hash_str &out) {
...
+		// and the final block is not correctly formatted.
+		int ret = EVP_DecryptFinal_ex(context, reinterpret_cast<unsigned char *>(out) + out_len,
+		                              reinterpret_cast<int *>(&out_len));
+		text_len += out_len;
+
+		if (ret > 0) {
+			// success
+			return text_len;
+
+		} else {
```
Comment (Mytherin) [inline extension/httpfs/crypto.cpp:140]: nit: this `else` is not necessary - generally we prefer early-out

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Nice results! Some comments from my side:

File: extension/httpfs/include/crypto.hpp (hunk 11 above)
Comment (Tishj) [reply extension/httpfs/include/crypto.hpp:51]: `make_shared_ptr` now (<https://github.com/duckdb/duckdb/pull/11696>)

File: src/include/duckdb/main/config.hpp (hunk 2 above)
Comment (ccfelius) [reply src/include/duckdb/main/config.hpp:107]: I think we can indeed do that. Will work on it!

File: extension/parquet/parquet_crypto.cpp (hunk 4 above)
Comment (ccfelius) [reply extension/parquet/parquet_crypto.cpp:156]: The different encryption algorithms by Parquet are  AES GCM and a combination of AES GCM and AES CTR, where in the latter case only data pages itself are encrypted with CTR but the other modules with GCM. Considering that they only support AES, we can maybe just rename it to AESState?

File: extension/httpfs/httpfs_extension.cpp (hunk 7 above)
Comment (carlopi) [reply extension/httpfs/httpfs_extension.cpp:65]: Very curious PR, expecially with the inter-extensions dependencies.

Would it not make sense to have the setting have 3 states, like:
* `SET parquet_encryption_mode = 'openssl';` -> needs HTTPFS extension loaded, errors out if no httpfs is there (possibly tries autoinstall + loading, maybe a follow up)
* `SET parquet_encryption_mode = 'mbedtls';` -> uses mbedtls backend
* `SET parquet_encryption_mode = 'auto';` -> uses openssl if available (but never changes set of extension loaded), otherwise mbedtls. This could be the default.

Comment (Mytherin) [reply extension/httpfs/httpfs_extension.cpp:65]: I think this should really only be a debug setting, so I would not think too much about how to set this up. It should actually also be prefixed with `debug` imo

File: extension/httpfs/include/crypto.hpp (hunk 9 above)
Comment (ccfelius) [reply extension/httpfs/include/crypto.hpp:43]: The reason I kept it like this is such that the header file `crypto.hpp` in httpfs extension is not dependent on OpenSSL libraries. But if that doesn't matter I'll change it to `EVP_CIPHER_CTX`

Comment (Mytherin) [reply extension/httpfs/include/crypto.hpp:43]: Since it is a pointer anyway, can you perhaps use a forward declaration?

File: src/include/duckdb/common/aes_state.hpp [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+#pragma once
+
+#include "duckdb/common/assert.hpp"
+#include "duckdb/common/helper.hpp"
+#include "duckdb/common/optional_ptr.hpp"
+#include "duckdb/common/shared_ptr.hpp"
+#include "duckdb/storage/storage_extension.hpp"
+
+// does this matter?
```
Comment (lnkuiper) [inline src/include/duckdb/common/aes_state.hpp:17]: Is the question whether the namespace matters? It does! Can this comment be removed?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Hi @ccfelius, thanks for the PR! It's amazing how little overhead there is when openssl is used.

Could you add a test where you encrypt a file using mbedtls, and then decrypt using openssl, and the other way around? This should work properly since the AES logic is the same.

File: src/include/duckdb/main/config.hpp (hunk 2 above)
Comment (ccfelius) [reply src/include/duckdb/main/config.hpp:107]: @Mytherin After looking into it it seems a bit tricky for ParquetWriter since it does not make use of the `parquet_options` and its not trivial (for me) how to pass these to the Parquet writer. For the ParquetReader it works however. Could you maybe give a pointer on how we can work around this? 

Or: maybe the whole option is not necessary. Removing `require httpfs` in test files or benchmark files is also sufficient to default to mbedtls.

File: extension/httpfs/httpfs_extension.cpp (hunk 7 above)
Comment (ccfelius) [reply extension/httpfs/httpfs_extension.cpp]: I changed it now to this to still enable mbedtls:

For copy:
```
# write files with OpenSSL disabled (if httpfs is loaded)
COPY (SELECT 42 i) to '__TEST_DIR__/encrypted${key_len}_openssl.parquet' 
(ENCRYPTION_CONFIG {footer_key: 'key${key_len}'}, 
DEBUG_USE_OPENSSL false)
```

For read:
```
# read encrypted files by using mbedtls (if httpfs is loaded)
SELECT * FROM read_parquet('__TEST_DIR__/encrypted${key_len}_openssl.parquet', 
encryption_config={footer_key: 'key${key_len}'}, 
debug_use_openssl=false)
```

File: extension/httpfs/include/crypto.hpp (hunk 11 above)
Comment (ccfelius) [reply extension/httpfs/include/crypto.hpp]: @Tishj I pulled your changes from the main branch, and now this PR is a bit messy. Sorry..

**Comment by ccfelius** [comment]
I now get a CI failure because I changed the `serialize_parquet.cpp`, which I actually don't understand why I'm getting it

**Comment by carlopi** [comment]
> I now get a CI failure because I changed the `serialize_parquet.cpp`, which I actually don't understand why I'm getting it

Hard to say what was failing (do you have a link?), but your changes to parquet serialization look off, in the sense that you should not change meaning or order of the IDs, given that allows files written with older version (say just before your change) to be read also after your change.
You likely should add an optional member (WritePropertyWithDefault is right) at the end of the other fields, without changing numbering of other ones.

File: extension/httpfs/crypto.cpp [hunk 15]
```diff
@@ -24,4 +30,118 @@ void hex256(hash_bytes &in, hash_str &out) {
...
+		throw InternalException("AES CTR failed with EncryptInit");
+	}
+}
+
+void AESGCMStateSSL::InitializeDecryption(const_data_ptr_t iv, idx_t iv_len, const string *key) {
+	auto context = reinterpret_cast<evp_cipher_ctx_st *>(gcm_context);
+	mode = DECRYPT;
+
+	if (1 != EVP_DecryptInit_ex(context, GetCipher(*key), NULL, const_data_ptr_cast(key->data()), iv)) {
+		throw std::runtime_error("EVP_EncryptInit_ex failed");
```
Comment (Tishj) [inline extension/httpfs/crypto.cpp:83]: Is there a reason this is a std::runtime_error instead of a DuckDB exception?

File: extension/httpfs/include/crypto.hpp [hunk 16]
```diff
@@ -16,4 +21,37 @@ void hmac256(std::string message, hash_bytes secret, hash_bytes &out);
...
+public:
+	DUCKDB_API bool IsOpenSSL() override;
+	DUCKDB_API void InitializeEncryption(const_data_ptr_t iv, idx_t iv_len, const std::string *key) override;
+	DUCKDB_API void InitializeDecryption(const_data_ptr_t iv, idx_t iv_len, const std::string *key) override;
+	DUCKDB_API size_t Process(const_data_ptr_t in, idx_t in_len, data_ptr_t out, idx_t out_len) override;
+	DUCKDB_API size_t Finalize(data_ptr_t out, idx_t out_len, data_ptr_t tag, idx_t tag_len) override;
+	DUCKDB_API void GenerateRandomData(data_ptr_t data, idx_t len) override;
+
+public:
+	static constexpr size_t BLOCK_SIZE = 16;
```
Comment (Tishj) [inline extension/httpfs/include/crypto.hpp:39]: Nit: I believe we prefer idx_t over size_t

File: extension/parquet/include/parquet.json [hunk 17]
```diff
@@ -79,6 +79,11 @@
         "name": "encryption_config",
         "type": "shared_ptr<ParquetEncryptionConfig>",
         "default": "nullptr"
+      },
+      {
+        "id": 105,
+        "name": "debug_use_openssl",
+        "type": "bool"
```
Comment (Tishj) [inline extension/parquet/include/parquet.json:86]: I believe this needs a default value for backwards compatibility reasons

File: extension/parquet/include/parquet_crypto.hpp [hunk 18]
```diff
@@ -71,17 +72,20 @@ class ParquetCrypto {
 
 public:
 	//! Decrypt and read a Thrift object from the transport protocol
-	static uint32_t Read(TBase &object, TProtocol &iprot, const string &key);
+	static uint32_t Read(TBase &object, TProtocol &iprot, const string &key, const shared_ptr<EncryptionState> &aes_p);
```
Comment (Tishj) [inline extension/parquet/include/parquet_crypto.hpp:75]: If this is const anyways is there a reason for this to be a reference to a shared_ptr instead of an EncryptionState, or an optional_ptr if it can be NULL?

File: extension/parquet/parquet_reader.cpp [hunk 19]
```diff
@@ -484,23 +479,33 @@ ParquetColumnDefinition ParquetColumnDefinition::FromSchemaValue(ClientContext &
...
 	file_handle = fs.OpenFile(file_name, FileFlags::FILE_FLAGS_READ);
 	if (!file_handle->CanSeek()) {
 		throw NotImplementedException(
 		    "Reading parquet files from a FIFO stream is not supported and cannot be efficiently supported since "
 		    "metadata is located at the end of the file. Write the stream to disk first and read from there instead.");
 	}
+
+	// set pointer to factory method for AES state
+	auto &config = DBConfig::GetConfig(context_p);
+	if (config.encryption_state && parquet_options.debug_use_openssl) {
```
Comment (Tishj) [inline extension/parquet/parquet_reader.cpp:493]: I wonder if we should use a new ClientContextState for this instead of putting this in the DBConfig?

File: extension/parquet/include/parquet_crypto.hpp (hunk 18 above)
Comment (ccfelius) [reply extension/parquet/include/parquet_crypto.hpp:75]: It can never be NULL here, so `optional_ptr` does not work. If I change everything to `unique_ptr`, I ultimately get the following error in `parquet_crypto.cpp`:

```Call to implicitly-deleted copy constructor of 'unique_ptr<duckdb::EncryptionState>``` 

So, to me, `shared_ptr` seemed the best solution

File: extension/parquet/parquet_reader.cpp (hunk 19 above)
Comment (ccfelius) [reply extension/parquet/parquet_reader.cpp:493]: Initially I thought about something like that too. But what happens when we have multiple clients? Will this then still work?

File: extension/httpfs/crypto.cpp (hunk 15 above)
Comment (ccfelius) [reply extension/httpfs/crypto.cpp:83]: No, thanks for pointing out!

File: extension/parquet/include/parquet_crypto.hpp (hunk 18 above)
Comment (Tishj) [reply extension/parquet/include/parquet_crypto.hpp:77]: Can't we change it to take a `const EncryptionState &` then?

Comment (ccfelius) [reply extension/parquet/include/parquet_crypto.hpp:77]: I've tested this, but the when I try to avoid the pointers it calls the abstract class - which has no implementation. `EncryptionState` can either contain a `AESGCMStateSSL` or a `AESGCMStateMBEDTLS` state, but which one it is we can only know at runtime. So I think what happens if I use  `const EncryptionState &` is that the states are automatically casted to the abstract class, which calls then only contains 'empty' functions.

Comment (Tishj) [reply extension/parquet/include/parquet_crypto.hpp:77]: The functions are virtual so that should work, that's also how the current `shared_ptr<EncryptionState> &` behaves, it's never cast to the concrete type right?
The only problem arises when you pass it by value instead of by references, then you get [object slicing](https://en.wikipedia.org/wiki/Object_slicing#:~:text=In%20C%2B%2B%20programming%2C%20object,%2C%20been%20%22sliced%20off%22)

File: extension/parquet/parquet_reader.cpp (hunk 19 above)
Comment (Tishj) [reply extension/parquet/parquet_reader.cpp:493]: Yea you're right, I believe the `encryption_state` is read-only though, so perhaps there could be an object in the HTTPFS extension that is only referenced by the ClientContextStates?

But that sounds needlessly brittle and error prone, I was just thinking out loud, feel free to disregard this comment :)

File: extension/parquet/include/parquet_crypto.hpp (hunk 18 above)
Comment (ccfelius) [reply extension/parquet/include/parquet_crypto.hpp:77]: Hmm yes I think you're right. Let me look into it again! Thanks :)

**Comment by lnkuiper** [comment]
Looks good! I'm fairly sure the failing Python test is unrelated to your changes, I've seen it fail spuriously before.

**Comment by hannes** [comment]
I wonder if it would be feasible to find the OpenSSL functions using `dlfun(dlopen(NULL), "name")` instead of this dance with the load. Often, we run in environments that have OpenSSL linked into them anyway.

**Comment by ccfelius** [comment]
> I wonder if it would be feasible to find the OpenSSL functions using `dlfun(dlopen(NULL), "name")` instead of this dance with the load. Often, we run in environments that have OpenSSL linked into them anyway.

Thanks for your suggestion! This sounds like a better alternative indeed. I'll look into it and try to make it work this way.

**Review by samansmink — COMMENTED** [review]
@ccfelius I think this is almost good to go. It looks great to me except one small issue, which is @Tishj's comment about the `const shared_ptr<EncryptionState> &aes_p` params.

I think all functions:
- `ParquetCrypto::Read`
- `ParquetCrypto::Write`
- `ParquetCrypto::ReadData`
- `ParquetCrypto::WriteData`
- `EncryptionTransport::EncryptionTransport`
- `DecryptionTransport::DecryptionTransport`

can take a reference to `EncryptionUtil` This way we prevent making unnecessary copies of shared_ptrs. Then in `DecryptionTransport::DecryptionTransport` and `EncryptionTransport::EncryptionTransport` you can just call the `CreateEncryptionState` to initialize the AES state.

**Comment by ccfelius** [comment]
> @ccfelius I think this is almost good to go. It looks great to me except one small issue, which is @Tishj's comment about the `const shared_ptr<EncryptionState> &aes_p` params.
> 
> I think all functions:
> 
> * `ParquetCrypto::Read`
> * `ParquetCrypto::Write`
> * `ParquetCrypto::ReadData`
> * `ParquetCrypto::WriteData`
> * `EncryptionTransport::EncryptionTransport`
> * `DecryptionTransport::DecryptionTransport`
> 
> can take a reference to `EncryptionUtil` This way we prevent making unnecessary copies of shared_ptrs. Then in `DecryptionTransport::DecryptionTransport` and `EncryptionTransport::EncryptionTransport` you can just call the `CreateEncryptionState` to initialize the AES state.

@samansmink Thanks for reviewing! I'll integrate this last change then!

**Comment by samansmink** [comment]
@Mytherin this is good from my side!

**Comment by Mytherin** [comment]
Thanks!
